File: 01-week-1--starter-algorithms/02-day-3--fibonacci-series/javascript/fibonacci.js
function fibonacci(num) {
  let [penult, ult, acc] = [0, 1, 0];

  for (let index = 0; index < num; index++) {
    penult = ult
    ult = acc
    acc = penult + ult;
  }

  return acc;
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log("=>", fibonacci(0));

  console.log("");

  console.log("Expecting: 1");
  console.log("=>", fibonacci(2));

  console.log("");

  console.log("Expecting: 8");
  console.log("=>", fibonacci(6));

  console.log("");

  console.log("Expecting: 55");
  console.log("=>", fibonacci(10));

  console.log("");

  console.log("Expecting: 196418");
  console.log("=>", fibonacci(27));

  console.log("");

  console.log("Expecting: 43466557686937456435688527675040625802564660517371780402481729089536555417949051890403879840079255169295922593080322634775209689623239873322471161642996440906533187938298969649928516003704476137795166849228875");
  console.log("=>", fibonacci(998));
}

module.exports = fibonacci;

// Please add your pseudocode to this file
/*
- Assign 0 to an accumulator (acc) variable
- Assign 1 to an ultimate (ult) variable
- Assign 0 to an penultimate (penult) variable
- Iterate a loop num times
- Each time, penult = ult, ult = acc, acc = penult + ult
- When the end of the loop is reached (or if loop doesn't iterate) return acc
*/

// And a written explanation of your solution
/*
Function takes in an index and returns the fib number at that index.

Keeps track of the second to last and last numbers so far in sequence and returns the accumulated value.
*/

File: 01-week-1--starter-algorithms/02-day-3--fibonacci-series/ruby/fibonacci.rb
def fibonacci(num)
  return num if num < 2

  last_pair = [0, 1]

  i = 0

  while i < num - 1
    sum = last_pair[0] + last_pair[1]
    last_pair = [last_pair[1], sum]
    i += 1
  end

  return last_pair[1]
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts "=>", fibonacci(0)

  puts

  puts "Expecting: 1"
  puts "=>", fibonacci(2)

  puts

  puts "Expecting: 55"
  puts "=>", fibonacci(10)

  # Don't forget to add your own!
  puts "Expecting: 196418"
  puts "=>", fibonacci(27)

  puts

  puts "Expecting: 196418"
  puts "=>", fibonacci(27)

  puts "Expecting: 317811"
  puts "=>", fibonacci(28)
end

# Please add your pseudocode to this file

# - If num is less than 2, return it.
# - Set an array containing 0 and 1.
# - Make a loop that runs num - 1 times
# - Each time, sum the numbers in array and set the array to contain the 2nd number as the first and the sum as the 2nd element.
# - Return the 2nd element.

# And a written explanation of your solution

# For the first two zero-indexed Fib numbers, the value is the same as the index so I return the index. For every index after that, the value is the sum of the last two values, so I loop through num - 1 times and return the final sum.
